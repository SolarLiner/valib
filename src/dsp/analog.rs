use nalgebra::Complex;
use crate::dsp::DSP;

pub trait DspAnalog<const I: usize, const O: usize>: DSP<I, O> {
    fn h_s(&self, s: [Complex<Self::Sample>; I]) -> [Complex<Self::Sample>; O];
}
