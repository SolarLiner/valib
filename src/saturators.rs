use num_traits::NumCast;
use numeric_literals::replace_float_literals;

use crate::Scalar;

pub trait Saturator<T: Scalar>: Default {
    /// Saturate an input with a frozen state.
    fn saturate(&self, x: T) -> T;

    /// Update the state given an input.
    #[inline(always)]
    fn update_state(&mut self, _x: T) {}

    #[inline(always)]
    #[replace_float_literals(T::from(literal).unwrap())]
    fn sat_diff(&self, x: T) -> T {
        (self.saturate(x + 1e-4) - self.saturate(x)) / 1e-4
    }
}

#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
pub struct Linear;

impl<S: Scalar> Saturator<S> for Linear {
    #[inline(always)]
    fn saturate(&self, x: S) -> S {
        x
    }

    #[inline(always)]
    fn sat_diff(&self, _: S) -> S {
        S::one()
    }
}

#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
pub struct Tanh;

impl<S: Scalar> Saturator<S> for Tanh {
    #[inline(always)]
    fn saturate(&self, x: S) -> S {
        x.tanh()
    }

    #[inline(always)]
    #[replace_float_literals(S::from(literal).unwrap())]
    fn sat_diff(&self, x: S) -> S {
        1. - x.tanh().powi(2)
    }
}

#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
pub struct Clipper;

impl<S: Scalar> Saturator<S> for Clipper {
    #[inline(always)]
    fn saturate(&self, x: S) -> S {
        x.min(S::one()).max(S::one().neg())
    }

    #[inline(always)]
    #[replace_float_literals(S::from(literal).unwrap())]
    fn sat_diff(&self, x: S) -> S {
        if x < -1. || x > 1. {
            0.
        } else {
            1.
        }
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct DiodeClipper<T> {
    param: T,
}

impl<T: Scalar> Default for DiodeClipper<T> {
    fn default() -> Self {
        // Parameter fit from a KiCad simulation of a buffer -> 2x3 diode clipper stage from -12V to 12V,
        // which yields a maximum difference of -20 dB.
        Self {
            param: T::from(0.75975449).unwrap(),
        }
    }
}

impl<T: Scalar> Saturator<T> for DiodeClipper<T> {
    #[inline(always)]
    fn saturate(&self, x: T) -> T {
        self.param * x / (self.param + x.abs())
    }

    #[inline(always)]
    fn sat_diff(&self, x: T) -> T {
        let frac = self.param / (self.param + x.abs());
        frac.powi(2)
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct Blend<T, S> {
    amt: T,
    inner: S,
}

impl<T: Scalar, S: Saturator<T>> Saturator<T> for Blend<T, S> {
    #[inline(always)]
    fn saturate(&self, x: T) -> T {
        x + self.amt * (self.inner.saturate(x) - x)
    }

    #[inline(always)]
    fn update_state(&mut self, x: T) {
        self.inner.update_state(x)
    }

    #[inline(always)]
    fn sat_diff(&self, x: T) -> T {
        T::one() + self.amt * (self.inner.sat_diff(x) - T::one())
    }
}

impl<T: NumCast, S: Default> Default for Blend<T, S> {
    fn default() -> Self {
        Self {
            amt: T::from(0.5).unwrap(),
            inner: S::default(),
        }
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum Dynamic<T> {
    Linear,
    Tanh,
    HardClipper,
    DiodeClipper(DiodeClipper<T>),
    SoftClipper(Blend<T, DiodeClipper<T>>),
}

impl<T: Scalar> Saturator<T> for Dynamic<T> {
    #[inline(always)]
    fn saturate(&self, x: T) -> T {
        match self {
            Self::Linear => Linear.saturate(x),
            Self::HardClipper => Clipper.saturate(x),
            Self::Tanh => Tanh.saturate(x),
            Self::DiodeClipper(clip) => clip.saturate(x),
            Self::SoftClipper(clip) => clip.saturate(x),
        }
    }

    #[inline(always)]
    fn sat_diff(&self, x: T) -> T {
        match self {
            Self::Linear => Linear.sat_diff(x),
            Self::HardClipper => Clipper.sat_diff(x),
            Self::Tanh => Tanh.sat_diff(x),
            Self::DiodeClipper(clip) => clip.sat_diff(x),
            Self::SoftClipper(clip) => clip.sat_diff(x),
        }
    }
}

impl<T> Default for Dynamic<T> {
    fn default() -> Self {
        Self::Linear
    }
}
