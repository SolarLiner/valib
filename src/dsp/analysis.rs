use nalgebra::{Complex, zero};
use simba::simd::SimdComplexField;
use numeric_literals::replace_float_literals;
use simba::scalar::RealField;
use crate::dsp::DSP;
use crate::Scalar;
use crate::dsp::analog::DspAnalog;

pub trait DspAnalysis<const I: usize, const O: usize>: DSP<I, O> {
    fn h_z(&self, z: [Complex<Self::Sample>; I]) -> [Complex<Self::Sample>; O];
    fn freq_response(&self, jw: [Self::Sample; I]) -> [Complex<Self::Sample>; O]
        where
            Self::Sample: RealField,
            Complex<Self::Sample>: SimdComplexField<SimdRealField=Self::Sample>
    {
        let z = jw.map(Complex::from_simd_real).map(|jw| <Complex<Self::Sample> as SimdComplexField>::simd_exp(Complex::i() * jw));
        self.h_z(z)
    }
}

impl<const I: usize, const O: usize, D: DspAnalog<I, O>> DspAnalysis<I, O> for D where Complex<D::Sample>: SimdComplexField {
    #[replace_float_literals(Complex::from(D::Sample::from_f64(literal)))]
    #[inline(always)]
    fn h_z(&self, z: [Complex<D::Sample>; I]) -> [Complex<D::Sample>; O] {
        self.h_s(z.map(|z| 2. * (z - 1.) / (z + 1.)))
    }

    #[inline(always)]
    fn freq_response(&self, jw: [D::Sample; I]) -> [Complex<D::Sample>; O]
        where
            Self::Sample: RealField,
    {
        self.h_s(jw.map(|jw| Complex::new(D::Sample::from_f64(0.0), jw)))
    }
}
