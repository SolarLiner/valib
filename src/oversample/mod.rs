use std::ops::{Deref, DerefMut};

use crate::dsp::DSP;
use crate::dsp::{
    buffer::{AudioBufferMut, AudioBufferRef},
    parameter::{HasParameters, Parameter},
};
use crate::saturators::Linear;
use crate::Scalar;
use crate::{
    dsp::{blocks::Series, DSPBlock},
    filters::biquad::Biquad,
};

#[derive(Debug, Clone)]
pub struct Oversample<T> {
    os_factor: usize,
    os_buffer: Box<[T]>,
    pre_filter: Series<[Biquad<T, Linear>; 8]>,
    post_filter: Series<[Biquad<T, Linear>; 8]>,
}

impl<T: Scalar> Oversample<T> {
    pub fn new(os_factor: usize, max_block_size: usize) -> Self {
        assert!(os_factor > 1);
        let os_buffer = vec![T::zero(); max_block_size * os_factor].into_boxed_slice();
        let filters = std::array::from_fn(|_| {
            Biquad::lowpass(
                T::from_f64(2.0 * os_factor as f64).simd_recip(),
                T::from_f64(0.707),
            )
        });
        Self {
            os_factor,
            os_buffer,
            pre_filter: Series(filters),
            post_filter: Series(filters),
        }
    }

    pub fn latency(&self) -> usize {
        2 * self.os_factor + DSP::latency(&self.pre_filter) + DSP::latency(&self.post_filter)
    }

    pub fn max_block_size(&self) -> usize {
        self.os_buffer.len() / self.os_factor
    }

    pub fn oversample(&mut self, buffer: &[T]) -> OversampleBlock<T> {
        let os_len = self.zero_stuff(buffer);
        for s in &mut self.os_buffer[..os_len] {
            *s = self.pre_filter.process([*s])[0];
        }
        OversampleBlock {
            filter: self,
            os_len,
        }
    }

    pub fn reset(&mut self) {
        self.os_buffer.fill(T::zero());
        DSP::reset(&mut self.pre_filter);
        DSP::reset(&mut self.post_filter);
    }

    pub fn with_dsp<P: DSPBlock<1, 1>>(self, dsp: P) -> Oversampled<T, P> {
        let max_block_size = dsp.max_block_size().unwrap_or(self.os_buffer.len());
        // Verify that we satisfy the inner DSPBlock instance's requirement on maximum block size
        assert!(self.os_buffer.len() <= max_block_size);
        let staging_buffer = vec![T::zero(); max_block_size].into_boxed_slice();
        Oversampled {
            oversampling: self,
            staging_buffer,
            inner: dsp,
        }
    }

    fn zero_stuff(&mut self, inp: &[T]) -> usize {
        let os_len = inp.len() * self.os_factor;
        assert!(self.os_buffer.len() >= os_len);

        self.os_buffer[..os_len].fill(T::zero());
        for (i, s) in inp.iter().copied().enumerate() {
            self.os_buffer[self.os_factor * i] = s * T::from_f64(self.os_factor as f64);
        }
        os_len
    }

    fn decimate(&mut self, out: &mut [T]) {
        let os_len = out.len() * self.os_factor;
        assert!(os_len <= self.os_buffer.len());

        for (i, s) in self
            .os_buffer
            .iter()
            .step_by(self.os_factor)
            .copied()
            .enumerate()
            .take(out.len())
        {
            out[i] = s;
        }
    }
}

pub struct OversampleBlock<'a, T> {
    filter: &'a mut Oversample<T>,
    os_len: usize,
}

impl<'a, T> Deref for OversampleBlock<'a, T> {
    type Target = [T];

    fn deref(&self) -> &Self::Target {
        &self.filter.os_buffer[..self.os_len]
    }
}

impl<'a, T> DerefMut for OversampleBlock<'a, T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.filter.os_buffer[..self.os_len]
    }
}

impl<'a, T: Scalar> OversampleBlock<'a, T> {
    pub fn finish(self, out: &mut [T]) {
        let filter = self.filter;
        for s in &mut filter.os_buffer[..self.os_len] {
            *s = filter.post_filter.process([*s])[0];
        }
        filter.decimate(out);
    }
}

pub struct Oversampled<T, P> {
    oversampling: Oversample<T>,
    staging_buffer: Box<[T]>,
    pub inner: P,
}

impl<T, P> Oversampled<T, P> {
    pub fn os_factor(&self) -> usize {
        self.oversampling.os_factor
    }
}

impl<T, P> Oversampled<T, P>
where
    T: Scalar,
{
    #[deprecated = "Use Oversample::with_dsp"]
    pub fn new(oversampling: Oversample<T>, inner: P) -> Self
    where
        P: DSP<1, 1, Sample = T>,
    {
        oversampling.with_dsp(inner)
    }
    pub fn into_inner(self) -> P {
        self.inner
    }
}

impl<T, P> DSPBlock<1, 1> for Oversampled<T, P>
where
    T: Scalar,
    P: DSPBlock<1, 1, Sample = T>,
{
    type Sample = T;

    fn latency(&self) -> usize {
        self.oversampling.latency() + self.inner.latency()
    }

    fn max_block_size(&self) -> Option<usize> {
        Some(self.oversampling.max_block_size())
    }

    fn reset(&mut self) {
        self.oversampling.reset();
        self.inner.reset();
    }

    fn process_block(&mut self, inputs: AudioBufferRef<T, 1>, mut outputs: AudioBufferMut<T, 1>) {
        let mut os_block = self.oversampling.oversample(inputs.get_channel(0));
        let input = &mut self.staging_buffer[..os_block.len()];
        input.copy_from_slice(&os_block);
        self.inner.process_block(
            AudioBufferRef::from(&*input),
            AudioBufferMut::from(&mut *os_block),
        );
        os_block.finish(outputs.get_channel_mut(0));
    }

    fn set_samplerate(&mut self, samplerate: f32) {
        self.inner
            .set_samplerate(self.oversampling.os_factor as f32 * samplerate);
    }
}

impl<S, P: HasParameters> HasParameters for Oversampled<S, P> {
    type Enum = P::Enum;

    fn get_parameter(&self, param: Self::Enum) -> &Parameter {
        self.inner.get_parameter(param)
    }
}

#[cfg(test)]
mod tests {
    use std::{f32::consts::TAU, hint::black_box};

    use numeric_literals::replace_float_literals;

    use crate::{
        dsp::{buffer::AudioBufferBox, DSPBlock as _},
        Scalar,
    };

    use super::Oversample;

    #[test]
    fn oversample_no_dc_offset() {
        let inp: [f32; 512] = std::array::from_fn(|i| (TAU * i as f32 / 64.).sin());
        let mut out = [0.0; 512];
        let mut os = Oversample::new(4, 512);

        let osblock = black_box(os.oversample(&inp));
        insta::assert_csv_snapshot!("os block", &*osblock, { "[]" => insta::rounded_redaction(3) });

        osblock.finish(&mut out);
        insta::assert_csv_snapshot!("post os", &out as &[_], { "[]" => insta::rounded_redaction(3) });
    }

    #[test]
    fn oversampled_dsp_block() {
        struct NaiveSquare<T> {
            samplerate: T,
            frequency: T,
            phase: T,
        }

        impl<T: Scalar> crate::dsp::DSP<1, 1> for NaiveSquare<T> {
            type Sample = T;

            #[replace_float_literals(T::from_f64(literal))]
            fn process(&mut self, _: [Self::Sample; 1]) -> [Self::Sample; 1] {
                let step = self.frequency / self.samplerate;
                let out = (1.0).select(self.phase.simd_gt(0.5), -1.0);
                self.phase += step;
                self.phase = (self.phase - 1.0).select(self.phase.simd_gt(1.0), self.phase);
                [out]
            }
        }

        let samplerate = 1000f32;
        let freq = 10f32;
        let dsp = NaiveSquare {
            samplerate,
            frequency: freq,
            phase: 0.0,
        };
        let mut os = Oversample::<f32>::new(4, 64).with_dsp(dsp);

        let input = AudioBufferBox::zeroed(64);
        let mut output = AudioBufferBox::zeroed(64);
        os.process_block(input.as_ref(), output.as_mut());
        insta::assert_csv_snapshot!(output.get_channel(0), { "[][]" => insta::rounded_redaction(3) });
    }
}
