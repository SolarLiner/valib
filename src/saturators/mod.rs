use crate::dsp::DSP;
use numeric_literals::replace_float_literals;

use crate::saturators::adaa::{Antiderivative, Antiderivative2};
use clippers::DiodeClipperModel;

use crate::Scalar;

pub mod adaa;
pub mod clippers;

#[allow(unused_variables)]
pub trait Saturator<T: Scalar>: Default {
    /// Saturate an input with a frozen state.
    fn saturate(&self, x: T) -> T;

    /// Update the state given an input and the output of `saturate`.
    #[inline(always)]
    fn update_state(&mut self, x: T, y: T) {}

    #[inline(always)]
    #[replace_float_literals(T::from_f64(literal))]
    fn sat_diff(&self, x: T) -> T {
        (self.saturate(x + 1e-4) - self.saturate(x)) / 1e-4
    }
}

#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
pub struct Linear;

impl<S: Scalar> Saturator<S> for Linear {
    #[inline(always)]
    fn saturate(&self, x: S) -> S {
        x
    }

    #[inline(always)]
    fn sat_diff(&self, _: S) -> S {
        S::one()
    }
}

#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
pub struct Tanh;

impl<S: Scalar> Saturator<S> for Tanh {
    #[inline(always)]
    fn saturate(&self, x: S) -> S {
        x.simd_tanh()
    }

    #[inline(always)]
    #[replace_float_literals(S::from_f64(literal))]
    fn sat_diff(&self, x: S) -> S {
        1. - x.simd_tanh().simd_powi(2)
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq, Default)]
pub struct Asinh;

impl<T: Scalar> Saturator<T> for Asinh {
    fn saturate(&self, x: T) -> T {
        x.simd_asinh()
    }

    fn sat_diff(&self, x: T) -> T {
        let x0 = x * x + T::one();
        x0.simd_sqrt().simd_recip()
    }
}

#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
pub struct Clipper;

impl<S: Scalar> Saturator<S> for Clipper {
    #[inline(always)]
    #[replace_float_literals(S::from_f64(literal))]
    fn saturate(&self, x: S) -> S {
        x.simd_min(1.0).simd_max(-1.0)
    }

    #[inline(always)]
    #[replace_float_literals(S::from_f64(literal))]
    fn sat_diff(&self, x: S) -> S {
        let mask = x.simd_abs().simd_gt(1.0);
        (1.0).select(mask, 0.0)
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct Blend<T, S> {
    amt: T,
    inner: S,
}

impl<T: Scalar, S: Saturator<T>> Saturator<T> for Blend<T, S> {
    #[inline(always)]
    fn saturate(&self, x: T) -> T {
        x + self.amt * (self.inner.saturate(x) - x)
    }

    #[inline(always)]
    fn update_state(&mut self, x: T, y: T) {
        self.inner.update_state(x, y)
    }

    #[inline(always)]
    fn sat_diff(&self, x: T) -> T {
        T::one() + self.amt * (self.inner.sat_diff(x) - T::one())
    }
}

impl<T: Scalar, S: Default> Default for Blend<T, S> {
    fn default() -> Self {
        Self {
            amt: T::from_f64(0.5),
            inner: S::default(),
        }
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum Dynamic<T> {
    Linear,
    Tanh,
    HardClipper,
    DiodeClipper(DiodeClipperModel<T>),
    SoftClipper(Blend<T, DiodeClipperModel<T>>),
}

impl<T: Scalar> Saturator<T> for Dynamic<T> {
    #[inline(always)]
    fn saturate(&self, x: T) -> T {
        match self {
            Self::Linear => Linear.saturate(x),
            Self::HardClipper => Clipper.saturate(x),
            Self::Tanh => Tanh.saturate(x),
            Self::DiodeClipper(clip) => clip.saturate(x),
            Self::SoftClipper(clip) => clip.saturate(x),
        }
    }

    #[inline(always)]
    fn sat_diff(&self, x: T) -> T {
        match self {
            Self::Linear => Linear.sat_diff(x),
            Self::HardClipper => Clipper.sat_diff(x),
            Self::Tanh => Tanh.sat_diff(x),
            Self::DiodeClipper(clip) => clip.sat_diff(x),
            Self::SoftClipper(clip) => clip.sat_diff(x),
        }
    }
}

impl<T> Default for Dynamic<T> {
    fn default() -> Self {
        Self::Linear
    }
}

#[derive(Debug, Clone, Copy)]
pub struct Slew<T> {
    pub max_diff: T,
    last_out: T,
}

impl<T: Scalar> Default for Slew<T> {
    fn default() -> Self {
        Self {
            max_diff: T::from_f64(1.0),
            last_out: T::from_f64(0.0),
        }
    }
}

impl<T: Scalar> DSP<1, 1> for Slew<T> {
    type Sample = T;

    fn latency(&self) -> usize {
        0
    }

    fn reset(&mut self) {
        self.last_out = T::from_f64(0.0);
    }

    fn process(&mut self, x: [Self::Sample; 1]) -> [Self::Sample; 1] {
        let y = self.slew(x[0]);
        self.last_out = y;
        [y]
    }
}

impl<T: Scalar> Slew<T> {
    pub fn new(samplerate: T, max_diff: T) -> Self {
        Self {
            max_diff: max_diff / samplerate,
            last_out: T::from_f64(0.0),
        }
    }

    pub fn set_max_diff(&mut self, max: T, samplerate: T) {
        self.max_diff = max / samplerate;
    }

    fn slew_diff(&self, x: T) -> T {
        let diff = x - self.last_out;
        diff.simd_clamp(-self.max_diff, self.max_diff)
    }

    fn slew(&self, x: T) -> T {
        self.last_out + self.slew_diff(x)
    }
}

impl<T: Scalar> Saturator<T> for Slew<T> {
    fn saturate(&self, x: T) -> T {
        self.slew(x)
    }

    fn sat_diff(&self, x: T) -> T {
        self.slew_diff(x)
    }

    fn update_state(&mut self, _x: T, y: T) {
        self.last_out = y;
    }
}
