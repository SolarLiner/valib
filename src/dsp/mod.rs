#![doc = include_str!("./README.md")]

use nalgebra::Complex;
use num_traits::Zero;

use crate::dsp::buffer::{AudioBufferBox, AudioBufferMut, AudioBufferRef};
use crate::Scalar;

use self::analysis::DspAnalysis;

pub mod analysis;
pub mod blocks;
pub mod buffer;
pub mod parameter;
pub mod utils;

/// Trait for interacting with a DSP algorithm, outside of processing. Shared by processors of both
/// per-sample algorithms and block-based algorithms.
#[allow(unused_variables)]
pub trait DSPMeta {
    /// Type of the audio sample used by this DSP instance.
    type Sample: Scalar;

    /// Sets the processing samplerate for this [`DSPProcess`] instance.
    fn set_samplerate(&mut self, samplerate: f32) {}

    /// Report the latency of this DSP instance, that is the time, in samples, it takes for an input sample to be
    /// output back.
    fn latency(&self) -> usize {
        0
    }

    /// Reset this instance. Parameters should be kept, but any memory and derived state should be put back to a
    /// well-known default value.
    fn reset(&mut self) {}
}

/// DSP trait. This is the main abstraction of the whole library.
///
/// Implementors of this trait are processes that work on a per-sample basis.
///
/// Multichannel I/O is supported, and is determined by the `I` and `O` const generics.
/// It's up to each implementor to document what the inputs and outputs mean.
/// There are no restrictions on the number of input or output channels in and of itself.
/// Implementors can support multiple configurations of I/O for different setups.
pub trait DSPProcess<const I: usize, const O: usize>: DSPMeta {
    /// Process a single sample of audio. No assumptions are made on the contents of said sample,
    /// so it can both work for working with audio data, but also control signals like frequency in Hertz
    /// for oscillators, or gate signals that are actually either 0 or 1.
    fn process(&mut self, x: [Self::Sample; I]) -> [Self::Sample; O];
}

/// Trait for DSP processes that take in buffers of audio instead of single-samples.
/// Documentation of [`DSPProcess`] still applies in here; only the process method changes.
pub trait DSPProcessBlock<const I: usize, const O: usize>: DSPMeta {
    /// Process a block of audio. Implementors should take the values of inputs and produce a stream into outputs,
    /// *as if* it was processed sample-by-sample. Ideally there should be no difference between a [`DSPProcess`] and [`DSPProcessBlock`]
    /// implementation, when both can exist. Note that [`DSPProcessBlock`] is blanket-implemented for all [`DSPProcess`] implementors,
    /// and as such, a simultaneous implementation of both onto one struct is impossible.
    ///
    /// Implementors should assume inputs and outputs are of the same length, as it is the caller's responsibility to make sure of that.
    fn process_block(
        &mut self,
        inputs: AudioBufferRef<Self::Sample, I>,
        outputs: AudioBufferMut<Self::Sample, O>,
    );

    /// Define an optional maximum buffer size alloed by this [`DSPProcessBlock`] instance. Callers into this instance must
    /// then only provide buffers that are up to this size in samples.
    #[inline(always)]
    fn max_block_size(&self) -> Option<usize> {
        None
    }
}

impl<P, const I: usize, const O: usize> DSPProcessBlock<I, O> for P
where
    P: DSPProcess<I, O>,
{
    #[inline(never)]
    fn process_block(
        &mut self,
        inputs: AudioBufferRef<P::Sample, I>,
        mut outputs: AudioBufferMut<P::Sample, O>,
    ) {
        if I == 0 && O == 0 {
            return;
        }
        for i in 0..inputs.samples() {
            outputs.set_frame(i, self.process(inputs.get_frame(i)))
        }
    }
}

/// Adapt a [`DSPProcessBlock`] instance to be able to used as a [`DSPProcess`].
///
/// This introduces as much latency as the internal buffer size is.
/// The internal buffer size is determined by either the max accepted buffer size of the inner instance, or is set
/// to 64 samples by default.
pub struct PerSampleBlockAdapter<P, const I: usize, const O: usize>
where
    P: DSPProcessBlock<I, O>,
{
    input_buffer: AudioBufferBox<P::Sample, I>,
    input_filled: usize,
    output_buffer: AudioBufferBox<P::Sample, O>,
    output_filled: usize,
    inner: P,
    pub buffer_size: usize,
}

impl<P, const I: usize, const O: usize> std::ops::Deref for PerSampleBlockAdapter<P, I, O>
where
    P: DSPProcessBlock<I, O>,
{
    type Target = P;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl<P, const I: usize, const O: usize> std::ops::DerefMut for PerSampleBlockAdapter<P, I, O>
where
    P: DSPProcessBlock<I, O>,
{
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.inner
    }
}

impl<P, const I: usize, const O: usize> PerSampleBlockAdapter<P, I, O>
where
    P: DSPProcessBlock<I, O>,
{
    pub const DEFAULT_BUFFER_SIZE: usize = 64;
    pub fn new(dsp_block: P) -> Self {
        Self::new_with_max_buffer_size(dsp_block, Self::DEFAULT_BUFFER_SIZE)
    }

    pub fn new_with_max_buffer_size(dsp_block: P, max_buffer_size: usize) -> Self {
        let buffer_size = dsp_block
            .max_block_size()
            .map(|mbs| mbs.min(max_buffer_size))
            .unwrap_or(max_buffer_size);
        Self {
            input_buffer: AudioBufferBox::zeroed(buffer_size),
            input_filled: 0,
            output_buffer: AudioBufferBox::zeroed(buffer_size),
            output_filled: buffer_size,
            buffer_size,
            inner: dsp_block,
        }
    }

    pub fn into_inner(self) -> P {
        self.inner
    }
}

impl<P, const I: usize, const O: usize> DSPMeta for PerSampleBlockAdapter<P, I, O>
where
    P: DSPProcessBlock<I, O>,
{
    type Sample = P::Sample;

    fn set_samplerate(&mut self, samplerate: f32) {
        self.inner.set_samplerate(samplerate);
    }

    fn latency(&self) -> usize {
        (self.inner.latency() + self.input_buffer.samples()).saturating_sub(1)
    }

    fn reset(&mut self) {
        self.input_filled = 0;
        self.output_filled = self.output_buffer.samples();
        self.input_buffer.fill(P::Sample::zero());
        self.output_buffer.fill(P::Sample::zero());
    }
}

impl<P, const I: usize, const O: usize> DSPProcess<I, O> for PerSampleBlockAdapter<P, I, O>
where
    P: DSPProcessBlock<I, O>,
{
    fn process(&mut self, x: [Self::Sample; I]) -> [Self::Sample; O] {
        self.input_buffer.set_frame(self.input_filled, x);
        self.input_filled += 1;
        if self.input_buffer.samples() == self.input_filled {
            self.inner
                .process_block(self.input_buffer.as_ref(), self.output_buffer.as_mut());
            self.input_filled = 0;
            self.output_filled = 0;
        }

        if self.output_filled < self.buffer_size {
            let ret = self.output_buffer.get_frame(self.output_filled);
            self.output_filled += 1;
            ret
        } else {
            [Self::Sample::zero(); O]
        }
    }
}

impl<P, const I: usize, const O: usize> DspAnalysis<I, O> for PerSampleBlockAdapter<P, I, O>
where
    P: DspAnalysis<I, O>,
{
    fn h_z(&self, z: Complex<Self::Sample>) -> [[Complex<Self::Sample>; O]; I] {
        self.inner.h_z(z)
    }
}

#[cfg(test)]
mod tests {
    use std::marker::PhantomData;

    use super::*;

    #[test]
    fn test_per_sample_block_adapter() {
        struct Counter<T>(PhantomData<T>);

        impl<T: Scalar> Counter<T> {
            pub fn new() -> Self {
                Self(PhantomData)
            }
        }

        impl<T: Scalar> DSPMeta for Counter<T> {
            type Sample = T;
        }

        impl<T: Scalar> DSPProcessBlock<0, 1> for Counter<T> {
            fn process_block(
                &mut self,
                inputs: AudioBufferRef<T, 0>,
                mut outputs: AudioBufferMut<T, 1>,
            ) {
                let len = inputs.samples();
                assert_eq!(len, outputs.samples());

                for (i, out) in outputs.get_channel_mut(0).iter_mut().enumerate() {
                    *out = T::from_f64(i as _);
                }
            }
        }

        let mut adaptor = PerSampleBlockAdapter::new_with_max_buffer_size(Counter::<f32>::new(), 4);
        assert_eq!(3, adaptor.latency());

        let expected = [0.0, 0.0, 0.0, 0.0, 1.0, 2.0, 3.0, 0.0];
        let mut actual = [0.0; 8];

        // Calling `process_block` but it's actually calling the impl for all `DSP` passing each sample through.
        adaptor.process_block(
            AudioBufferRef::empty(8),
            AudioBufferMut::new([&mut actual]).unwrap(),
        );

        assert_eq!(expected, actual);
    }
}
