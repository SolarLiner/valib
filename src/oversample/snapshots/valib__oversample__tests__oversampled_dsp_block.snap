---
source: src/oversample/mod.rs
expression: "&output as &[_]"
---
-0.0
-0.0
-0.0
-0.0
-0.001
-0.015
-0.133
-0.52
-1.014
-1.135
-0.996
-0.976
-1.006
-1.002
-0.999
-1.0
-1.0
-0.997
-0.943
-0.599
0.306
1.18
1.209
0.951
0.971
1.014
1.0
0.999
1.0
0.999
0.984
0.832
0.208
-0.821
-1.296
-1.054
-0.936
-1.004
-1.007
-0.997
-1.0
-1.0
-0.997
-0.943
-0.599
0.306
1.18
1.209
0.951
0.971
1.014
1.0
0.999
1.0
0.999
0.984
0.832
0.208
-0.821
-1.296
-1.054
-0.936
-1.004
-1.007
